show closed action message when a workspace member is removed

marcochavezf opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
This PR shows the message closed action message ("This workspace chat is no longer active because...") in the LHN and the Search view when a member is removed from the workspace.

### Fixed Issues
<!---
Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing.
Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the issue; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/Expensify/issues/204344

### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Create a new workspace from an account that's on the `policyExpenseChat` beta (e.g @expensifail.com)
2. Add a new member to the workspace via Manage Members > Invite 
3. Log-in as the member and send a message in the member's workspace chat
4. Log back in again to the admin account and remove the member from the workspace via Manage Members > Remove 
5. Verify the row chat (in LHN and search bar) should display the message "This workspace chat is no longer active because..."
6. Now log-in as the member again
7. Verify the row chat (in LHN and search bar) includes the same message from step 5.

- [x] Verify that no errors appear in the JS console

### PR Review Checklist
<!--
This is a checklist for PR authors & reviewers. Please make sure to complete all tasks and check them off once you do, or else Expensify has the right not to merge your PR!
-->
#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.


<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
<!-- Insert screenshots of your changes on the web platform-->
<img width="900" alt="Screen Shot 2022-04-25 at 18 07 26" src="https://user-images.githubusercontent.com/6829422/165189125-1aaf8273-3244-4cb0-84e8-7dc07ae62aac.png">

#### Mobile Web
<!-- Insert screenshots of your changes on the web platform (from a mobile browser)-->
<img width="329" alt="Screen Shot 2022-04-25 at 18 41 18" src="https://user-images.githubusercontent.com/6829422/165191866-fec62872-3799-49ad-a708-382ab53ef0ea.png">

#### Desktop
<!-- Insert screenshots of your changes on the desktop platform-->
<img width="1202" alt="Screen Shot 2022-04-25 at 18 40 04" src="https://user-images.githubusercontent.com/6829422/165198128-c4adb554-cfb9-47a5-97b5-afa0a329acd2.png">

#### iOS
<!-- Insert screenshots of your changes on the iOS platform-->
<img width="325" alt="Screen Shot 2022-04-25 at 19 57 59" src="https://user-images.githubusercontent.com/6829422/165198014-ab6bf6f9-c344-4d65-9ca2-a973306fdeb2.png">

#### Android
<!-- Insert screenshots of your changes on the Android platform-->
<img width="426" alt="Screen Shot 2022-04-25 at 18 51 37" src="https://user-images.githubusercontent.com/6829422/165192627-84f84f4f-a483-4b48-b10b-7f00fcc81620.png">

marcochavezf committed:
pass lastActionType property

marcochavezf committed:
add getLastAction to ReportActions

marcochavezf committed:
get last message according to last action

marcochavezf committed:
remove unnessary lastActionType

marcochavezf committed:
clean up a little bit the logic to get last message

marcochavezf committed:
fix warning message with SubcriptAvatar secondary icon

marcochavezf commented on src/components/OptionRow.js:
```diff
@@ -163,7 +163,7 @@ const OptionRow = (props) => {
                                 !_.isEmpty(props.option.icons)
                                 && (
-                                    props.option.shouldShowSubscript ? (
+                                    props.option.shouldShowSubscript && props.option.icons.length > 1 ? (
```
This condition was added to prevent a warning message in the console, where the SubscriptAvatar is expecting 2 icons, but when the workspace is archived, there's only one (trash) icon.

tgolen replied:
Hm, if the subscript avatar is always expecting two icons, then shouldn't the conditional be `.length === 2`?

tgolen replied:
and isn't it weird how that logic kind of defeats the purpose of `shouldShowSubscript`?

tgolen replied:
It seems to me that when the workspace is archived, then `shouldShowSubscript` should be `false` then?

tgolen commented on src/libs/actions/ReportActions.js:
```diff
@@ -111,9 +111,14 @@ function getLastVisibleMessageText(reportID) {
     );
 }
 
+function getLastAction(reportID) {
```
JSdocs are required when params exist.

tgolen replied:
I'm also not too big of a fan of this method. I think it would be better to have `OptionsListUtils` connect to Onyx to keep track of the last report action. Action files should be for the UI to trigger API requests and manage where the Onyx data goes. Using Action files for getters is kind of an anti-pattern.

marcochavezf replied:
Removed that function and added a `reportActions` subscription in `OptionsListUtils` to get the last action of a report.

marcochavezf committed:
don't show SubscriptAvatar when workspace is archived

marcochavezf committed:
add reportActions subscription in OptionListUtils

marcochavezf committed:
remove getLastAction from ReportActions

marcochavezf committed:
use reportActions object to get archiveReason

marcochavezf commented:
Updated `shouldShowSubscript` to be false if the workspace chat is archived (instead of checking the icons.length) and added a `reportActions` subscription in `OptionListUtils` in order to remove the getter function from `ReportActions`.

tgolen commented on src/libs/OptionsListUtils.js:
```diff
@@ -76,6 +77,18 @@ Onyx.connect({
+            return;
+        }
+        const reportID = CollectionUtils.extractCollectionItemID(key);
+        reportActions[reportID] = _.toArray(actions);
```
Since this only needs to keep track of the last report action, how about this? It'll require just a few minor changes below.

```suggestion
        lastReportAction[reportID] = _.last(_.toArray(actions));
```

marcochavezf replied:
Sounds good! The code looks a bit cleaner below and the `_.last` calculation is made just once

marcochavezf replied:
One minor thing though, I changed it to `lastReportActions` (plural) because it's a collection, is that ok? Or should it be singular?

tgolen replied:
Yeah, plural is fine.

tgolen reviewed (changes requested):
Ah, that looks better, thanks!

marcochavezf committed:
change reportActions to lastReportActions

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Cherry-picked](https://github.com/Expensify/App/actions/runs/2296148142) to staging by @sketchydroide in version: 1.1.57-8 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

@Expensify/applauseleads please QA this PR and check it off on the [deploy checklist](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3AStagingDeployCash) if it passes.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/2320473434) to production by @chiragsalian in version: 1.1.57-17 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
